### ISSUE TITLE
bug: ibis duckdb insert does not handle column case properly

### ISSUE THREAD

**Issue by ruiyang2015** [ibis-project/ibis#8045]
### What happened?

when i try to user ibis.duckdb backend to insert a table expression to an existing table, the column case is not properly handled, here is a repo script:
```
import ibis
import pandas as pd
df1 = pd.DataFrame([[1],[2],[3],[4]], columns=['FTHG'])
df2 = pd.DataFrame([[5],[6],[7],[8]], columns=['FTHG'])

c = ibis.duckdb.connect()
c.create_table('t1', df1)
t2 = c.create_table('t2', df2)
c.insert('t1', t2) # --> an exception is thrown here

```

### What version of ibis are you using?

7.2.0

### What backend(s) are you using, if any?

Duckdb

### Relevant log output

```sh
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/ibis/backends/base/sql/alchemy/__init__.py", line 770, in insert
    bind.execute(to_table.insert().from_select(columns, compiled))
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1385, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 334, in _execute_on_connection
    return connection._execute_clauseelement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1569, in _execute_clauseelement
    compiled_sql, extracted_params, cache_hit = elem._compile_w_cache(
                                                ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 545, in _compile_w_cache
    compiled_sql = self._compiler(
                   ^^^^^^^^^^^^^^^
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 567, in _compiler
    return dialect.statement_compiler(dialect, self, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py", line 809, in __init__
    Compiled.__init__(self, dialect, statement, **kwargs)
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py", line 464, in __init__
    self.string = self.process(self.statement, **compile_kwargs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py", line 499, in process
    return obj._compiler_dispatch(self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
    return meth(self, **kw)
           ^^^^^^^^^^^^^^^^
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py", line 4015, in visit_insert
    crud_params = crud._get_crud_params(
                  ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/sql/crud.py", line 139, in _get_crud_params
    _scan_insert_from_select_cols(
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/sql/crud.py", line 320, in _scan_insert_from_select_cols
    cols = [stmt.table.c[_column_as_key(name)] for name in stmt._select_names]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/sql/crud.py", line 320, in <listcomp>
    cols = [stmt.table.c[_column_as_key(name)] for name in stmt._select_names]
            ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ruiyang/Go/src/github.com/ascend-io/ascend-core/.venv/lib/python3.11/site-packages/sqlalchemy/sql/base.py", line 1214, in __getitem__
    return self._index[key]
           ~~~~~~~~~~~^^^^^
KeyError: 'fthg'
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by cpcloud** [comment]
Thanks for the issue.

This is ultimately caused by SQLAlchemy allowing dialects to second-guess users' column case and try to normalize it in some way in an attempt to allow database identifiers to be case-insensitive while also following case sensitivity in Python.

In my experience, no dialect seems to get this completely right and it's extremely confusing for users when something goes awry.

AFAICT the most robust approach is to quote everything so that the database always preserves the case of every identifier.

This issue isn't present on a refactor we're doing right now in the `upstream/the-epic-split` branch. I'll create a PR to add a test for this case there, but I'm not sure we'll fix this until 9.0.0 (landing hopefully around mid February/early March 2024).

If someone else wants to pick this up in the meantime and put in the effort to address this for 8.0.0, please do.

**Comment by jcrist** [comment]
Since this is fixed in `upstream/the-epic-split` and already has a test in that branch, I'm going to close this for now.
